File: Backend/app/blueprint/metodos/gaussian_elimination_no_pivot.py
import numpy as np


def gaussian_elimination_no_pivoting(A, b):
    """
    Solves the system of linear equations Ax = b using Gaussian Elimination without Pivoting.

    Parameters:
    A (list of lists or np.ndarray): Coefficient matrix.
    b (list or np.ndarray): Right-hand side vector.

    Returns:
    np.ndarray: Solution vector x.
    """
    # Convert A and b into numpy arrays
    A = np.array(A, float)
    b = np.array(b, float)
    n = len(b)

    # Augment A with b to form the augmented matrix
    augmented_matrix = np.hstack([A, b.reshape(-1, 1)])

    # Forward elimination (without pivoting)
    for i in range(n):
        # Check if the diagonal element is zero, which would lead to division by zero
        if augmented_matrix[i, i] == 0:
            raise ValueError(f"Zero pivot encountered at row {i}. No pivoting applied.")

        # Eliminate entries below the pivot
        for j in range(i + 1, n):
            factor = augmented_matrix[j, i] / augmented_matrix[i, i]
            augmented_matrix[j, i:] -= factor * augmented_matrix[i, i:]

    # Back substitution to solve for x
    x = np.zeros(n)
    for i in range(n - 1, -1, -1):
        x[i] = (augmented_matrix[i, -1] - np.dot(augmented_matrix[i, i + 1:], x[i + 1:])) / augmented_matrix[i, i]

    return x


# Example usage:
A = [[2, -1, 1],
     [3, 3, 9],
     [3, 3, 5]]

b = [8, -6, -4]

solution = gaussian_elimination_no_pivoting(A, b)
print("Solution:", solution)

File: rust_methods/gaussian_elimination_no_pivot.rs
fn gaussian_elimination_no_pivoting(a: &mut Vec<Vec<f64>>, b: &mut Vec<f64>) -> Vec<f64> {
    let n = b.len();

    // Forward elimination (without pivoting)
    for i in 0..n {
        // Check if the diagonal element is zero, which would lead to division by zero
        if a[i][i] == 0.0 {
            panic!("Zero pivot encountered at row {}. No pivoting applied.", i);
        }

        // Eliminate entries below the pivot
        for j in i + 1..n {
            let factor = a[j][i] / a[i][i];
            for k in i..n {
                a[j][k] -= factor * a[i][k];
            }
            b[j] -= factor * b[i];
        }
    }

    // Back substitution to solve for x
    let mut x = vec![0.0; n];
    for i in (0..n).rev() {
        x[i] = b[i];
        for j in i + 1..n {
            x[i] -= a[i][j] * x[j];
        }
        x[i] /= a[i][i];
    }

    x
}

fn main() {
    // Example usage:
    let mut a = vec![
        vec![2.0, 1.0, -1.0],
        vec![-3.0, -1.0, 2.0],
        vec![-2.0, 1.0, 2.0],
    ];
    let mut b = vec![8.0, -11.0, -3.0];

    let solution = gaussian_elimination_no_pivoting(&mut a, &mut b);
    println!("Solution: {:?}", solution);
}
